### PR TITLE
Cam's SD Card Work

### PR DESCRIPTION
**Pull request by CClarke513**
SD Card preliminary work

### CODE REVIEW

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: I tend to use all-caps for constants, so these could be `CHIP_SELECT` or something like that. Not a huge concern either way.

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: ...equivalent to what?

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: The `sd_` in the prefix isn't necessary. This is the SDCard class, so it'll already be e.g. `sdCard.initialize()`. `sdCard.sd_initialize()` is just redundant.

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: We probably don't want to lock up the whole board due to an SD error, do we? I see at least two major problems with this:

- If we don't insert the SD card until after this point is reached, it never checks again.
- With the SD card absent, we can still send data over the radio.

Notice what I do with the sensors:

- Each sensor has a `getStatus()` method that tells you whether it's ready.
- The loop checks `getStatus()` before using the sensor, and if the sensor isn't connected, it tries to initialize it again.

This way you don't accidentally attempt to write to a nonexistant SD card, but also if it isn't there it isn't the end of the world.

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: Oh, good idea, I didn't even think of that. Do we want to try `open` again if not?

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h:18]: This is partly a design-level question that warrants more discussion, so I'll ping @DarthSethious1 as well.

The GPS updates far less frequently than the other sensors. For the radio, to keep the frame shape consistent, we accepted that sometimes the packet will have outdated GPS data. However, we don't have that restriction here: we can leave GPS values blank instead of putting older data. Which do we want to do?

If we decide to put older data so that every row has something, nothing needs to change here. If we decide to put blanks where we don't have up-to-date data, that would necessitate one of two changes:

- Add a second version of the function without this parameter
- Pass it by pointer instead (`const GPS::Coordinates* coordsPtr`) and pass `nullptr` when no new data is available

I don't have a preference for which route we take.

Comment (DarthSethious1) [reply SD_Card.h:18]: I think keeping this consistent between radio and SD card as often as possible is good. I'm fine with us writing old GPS values to SD Card and only updating it a fraction of the frames.

File: SD_Card.h
Comment (DarthSethious1) [inline SD_Card.h]: Since you're including values for the VOC, temp, and humidity, I will assume this code will work for Capsule 2, but not for Capsule 1. @bbrk24 are we able to write SD card code that is able to accommodate both capsules in the same file? Ie, is there something that can be done here to make this general enough for both capsules?

Comment (bbrk24) [reply SD_Card.h]: The fix is to simply wrap the Capsule 2-specific parts of the code in `#if CAPSULE == 2`. I defined that variable specifically to make things like this easier. I already do this in parts of frame.h -- though I forgot to do it for the parameter list to `createFrame` at first, I fixed that in #5.

Comment (DarthSethious1) [reply SD_Card.h]: awesome, Cam please do this for the SD card write code as well

File: SD_Card.h
Comment (bbrk24) [reply SD_Card.h]: > for constants

So not the `File`.

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: Same naming issue here with `sd_`

File: SD_Card.h
Comment (CClarke513) [reply SD_Card.h]: Just fixed those changes - can you take a look, please?

**Review by bbrk24 — CHANGES_REQUESTED** [review]
You might want to double-check that last commit...

image

**Comment by CClarke513** [comment]
> You might want to double-check that last commit...
> 
> image

We can discuss this in class today. I do have some time-urgent homework I need to get on for robotics!

File: SD_Card.h
Comment (bbrk24) [reply SD_Card.h]: Looks like you did it correctly in the body of the function, but the parameters should also be `#if`d out: https://github.com/bbrk24/BSC_Arduino/blob/392f541522fee0b189430039aa29906674bbd708/frame.h#L87-L93

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: hmm...

**Review by bbrk24 — CHANGES_REQUESTED** [review]

File: SD_Card.h
Comment (bbrk24) [reply SD_Card.h]: @TheRoyalIrishman I don't feel like this was properly addressed. Removing the `SD.begin` call is a non-starter as now every read and write will fail -- the SPI communications were never set up. My suggestion is to look at how the altimeter, IMU, or humidity sensor's `initialize` and `getStatus` methods are implemented (the GPS is a bit different because it has a multi-phase initialization).

**Review by DarthSethious1 — APPROVED** [review]
The problems I saw from the first draft were resolved, this code seems good to me

File: SD_Card.h
Comment (bbrk24) [inline SD_Card.h]: I won't approve this unless you at least put the `SD.begin()` call back. We can do the status check in a later PR because we need to configure the LED separately anyways

**Review by bbrk24 — CHANGES_REQUESTED** [review]

**Review by bbrk24 — APPROVED** [review]

**Pull request merged**
